### PR TITLE
1.18 branch

### PR DESCRIPTION
**Pull request by cooljeanius**
WIP

### CODE REVIEW

File: lua/common.lua [hunk 1]
```diff
@@ -42,3 +41,72 @@ end
...
+function wesnoth.wml_actions.remove_array_duplicates(cfg)
+	local name = cfg.name
+	local attribute = cfg.attribute
+
+	local values = {}
+	local inArray = wml.array_access.get(name)
+	local outArray = {}
+	for _, u in pairs(inArray) do
+		if values[u[attribute]] then
+			-- skip duplicate
```
Comment (cooljeanius) [inline lua/common.lua]: I think the `luacheck` warning here is legitimate; can't there at least be *something* that goes here?

**Comment by cooljeanius** [comment]
So, 2 possible options for the image optimization check:
- try optimizing them more manually, or
- increase the allowed threshold in the workflow from 45 to 55

I'm going to try the first one first...

File: lua/common.lua (hunk 1 above)
Comment (MadMax3947) [reply lua/common.lua]: Reversed the logic here which should resolve this.

Comment (cooljeanius) [reply lua/common.lua]: ...ok, I see this is resolved in [f302866](https://github.com/cooljeanius/Flight_Freedom/pull/22/commits/f3028668141506dff6811cc60220072059aaf982).

**Comment by MadMax3947** [comment]
> So, 2 possible options for the image optimization check:
> 
>     * try optimizing them more manually, or
> 
>     * increase the allowed threshold in the workflow from 45 to 55
> 
> 
> I'm going to try the first one first...

Removed EXIF data from a couple of the PNGs which should bring these images within the optimization threshold.

**Comment by cooljeanius** [comment]
> > So, 2 possible options for the image optimization check:
> > ```
> > * try optimizing them more manually, or
> > 
> > * increase the allowed threshold in the workflow from 45 to 55
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm going to try the first one first...
> 
> Removed EXIF data from a couple of the PNGs which should bring these images within the optimization threshold.

Can you add me as a collaborator on your fork so that I can push to this PR? I forgot to check the "Allow edits from maintainers" box when creating this PR...

**Comment by cooljeanius** [comment]
ok, so checks are all green now... I'll still want to give the changes a bit more of a manual look-over before merging...

File: lua/common.lua [hunk 2]
```diff
@@ -42,3 +41,70 @@ end
...
 	wesnoth.interface.clear_chat_messages()
 end
+
+---
+-- Get current game zoom level.
+--
+-- [get_zoom]
+--     variable=zoom
+-- [/get_zoom]
+---
```
Comment (cooljeanius) [inline lua/common.lua:51]: Just a note here that this might be useful to submit upstream for wesnoth/wesnoth#7423, but that's a separate topic...

**Comment by cooljeanius** [comment]
Attaching a zipfile of test saves for B branch testing: [test_saves.zip](https://github.com/user-attachments/files/15746075/test_saves.zip)

**Comment by cooljeanius** [comment]
looking through the attached saves, I see that there's one for each of the 7 playable scenarios in the B branch, which I'll take to imply adequate testing coverage, which ought to be enough for me to merge this

**Comment by cooljeanius** [comment]
Oh wait, I meant to get #24 done before merging this... oh well...

**Pull request merged**